### PR TITLE
feat: Send app data size in app context

### LINKED ISSUE #1664: Add how much space an app takes up on the user's device
### Problem Statement

In the iOS settings (General -> Storage) I can see how much space an app takes up. This is split into:

- App Size
- Documents & Data

The use case is, that I can see if the app creates too much data or if I don't properly clean up data.

Relates to https://github.com/getsentry/sentry-java/issues/1917

### Solution Brainstorm

_No response_

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

Calculate the size of the home directory (which is what "Documents & Data" is), in the background, every minute. And send it as part of the `app` context.

Screen Shot 2022-09-20 at 15 35 27

## :bulb: Motivation and Context

Closes #1664

## :green_heart: How did you test it?

WIP

## :pencil: Checklist

- [ ] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/include/SentryAppDataSizeObserver.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * A class that periodically reads the size of the Documents & Data
+ * folder, and stores it in `appDataSize`.
+ * It only does this on iOS, as on macOS it would read the entire user's
+ * home directory size, which is not helpful.
+ */
+@interface SentryAppDataSizeObserver : NSObject
+
+// The size of Documents & Data. This gets updated periodically.
+// A value of -1 means the size wasn't calculated (yet).
+@property (nonatomic) long long appDataSize;
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/include/SentryAppDataSizeObserver.h:15]: None of the other variables, like for memory and total and free storage, have something in their name. Would be kinda weird to only have it here, when literally all of them are in bytes?

File: Sources/Sentry/SentryClient.m [hunk 2]
```diff
@@ -816,6 +827,7 @@ - (void)removeExtraDeviceContextFromEvent:(SentryEvent *)event
                     key:@"app"
                   block:^(NSMutableDictionary *app) {
                       [app removeObjectForKey:SentryDeviceContextAppMemoryKey];
+                      [app removeObjectForKey:@"data_storage"];
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryClient.m:830]: As far as I can see, it doesn't exist for other platforms yet, it is undocumented. I said the same in the ticket.

File: Sources/Sentry/SentryAppDataSizeObserver.m [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+- (void)updateAppDataSize
+{
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0),
+        ^{ self.appDataSize = [self readAppDataSize]; });
+}
+
+// Code from https://github.com/NikolaiRuhe/NRFoundation/blob/master/NRFoundation/NRFileManager.m
+- (long long)readAppDataSize
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryAppDataSizeObserver.m:41]: I validated it by running the code and comparing the value to the actual folder that it checked. And this function was linked from the issue itself. Of course there is going to be overhead, that is why we do it in a background thread, once a minute.

As far as I can see from reading the docs, you can't get it much simpler than this: we get the contents of the folder (recursively), pre-fetch things like the file size, and fall back to getting the size manually if needed.

[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryAppDataSizeObserver.m:41]: If the overhead is significant, we should consider making this optional or not doing it.

File: Sources/Sentry/include/SentryAppDataSizeObserver.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/include/SentryAppDataSizeObserver.h:15]: Let's fix this in another PR.

File: Sources/Sentry/SentryAppDataSizeObserver.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryAppDataSizeObserver.m:41]: Even though it's in a background thread already? Overhead of course gets bigger the more files the app stores, but at that point that is probably exactly the info the developer wants to have.

We could add an option for it in SentryOptions, default it to true, but I honestly don't see the downside of doing this once a minute in a background thread.

**Comment by brustolin** [comment]
We could use a file observer to monitor file changes. Here is a sample https://github.com/tblank555/iMonitorMyFiles/blob/master/iMonitorMyFiles/Classes/TABFileMonitor.m

**Pull request closed without merging**